File: algorithms/0703.kth-largest-element-in-a-stream/solution.rb
class KthLargest

=begin
    :type k: Integer
    :type nums: Integer[]
=end
  def initialize(k, nums)
    @size = k
    @elements = nums.sort { |a, b| b <=> a }[0...k]
  end

=begin
    :type val: Integer
    :rtype: Integer
=end
  def add(val)
    index = @elements.bsearch_index { |x| x < val }
    if index
      @elements.insert(index, val)
    else
      @elements.push(val)
    end

    @elements.pop if @elements.size > @size

    @elements.last
  end
end

# Your KthLargest object will be instantiated and called as such:
# obj = KthLargest.new(k, nums)
# param_1 = obj.add(val)

File: algorithms/0703.kth-largest-element-in-a-stream/solution.rs
use std::cmp::Reverse;
use std::collections::BinaryHeap;

struct KthLargest {
    size: usize,
    heap: BinaryHeap<Reverse<i32>>,
}

/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl KthLargest {

    fn new(k: i32, nums: Vec<i32>) -> Self {
        let mut _self = KthLargest {
            size: k as usize,
            heap: BinaryHeap::new(),
        };
        for n in nums {
            _self.add(n);
        }
        _self
    }

    fn add(&mut self, val: i32) -> i32 {
        self.heap.push(Reverse(val));
        if self.heap.len() > self.size {
            self.heap.pop();
        }
        if let Some(Reverse(x)) = self.heap.peek() {
            *x
        } else {
            unreachable!()
        }
    }
}

/**
 * Your KthLargest object will be instantiated and called as such:
 * let obj = KthLargest::new(k, nums);
 * let ret_1: i32 = obj.add(val);
 */
